File: taurus-protect-sdk-typescript/src/models/reservation.ts
/**
 * Reservation models for Taurus-PROTECT SDK.
 *
 * Reservations are used to lock specific UTXOs (Unspent Transaction Outputs)
 * for UTXO-based blockchains like Bitcoin and Litecoin, preventing
 * double-spending during transaction creation.
 */

import type { Currency } from './currency';

/**
 * Represents a UTXO reservation in the Taurus-PROTECT system.
 *
 * Reservations lock specific UTXOs to prevent double-spending during
 * transaction creation on UTXO-based blockchains.
 */
export interface Reservation {
  /** Unique reservation identifier */
  readonly id?: string;
  /** Reserved amount in the currency's smallest unit */
  readonly amount?: string;
  /** Date when the reservation was created */
  readonly creationDate?: Date;
  /** The kind/type of reservation */
  readonly kind?: string;
  /** Optional comment about the reservation */
  readonly comment?: string;
  /** Internal address ID in Taurus-PROTECT */
  readonly addressId?: string;
  /** Blockchain address string/hash */
  readonly address?: string;
  /** Currency information for the reserved funds */
  readonly currencyInfo?: Currency;
  /** Resource ID associated with the reservation */
  readonly resourceId?: string;
  /** Resource type associated with the reservation */
  readonly resourceType?: string;
}

/**
 * Represents UTXO details for a reservation.
 *
 * Contains the UTXO (Unspent Transaction Output) data associated with a reservation.
 */
export interface ReservationUtxo {
  /** Unique UTXO identifier */
  readonly id?: string;
  /** Transaction hash */
  readonly hash?: string;
  /** Output index in the transaction */
  readonly outputIndex?: number;
  /** Locking script */
  readonly script?: string;
  /** Value in the smallest unit */
  readonly value?: string;
  /** Block height where the UTXO was created */
  readonly blockHeight?: string;
  /**
   * Request ID that reserved this UTXO.
   * @deprecated Replaced by reservationId since version 3.22
   */
  readonly reservedByRequestId?: string;
  /** Reservation ID */
  readonly reservationId?: string;
  /** Human-readable value string */
  readonly valueString?: string;
}

/**
 * Options for listing reservations.
 */
export interface ListReservationsOptions {
  /**
   * Filter by kind of reservation.
   * @deprecated Since 3.20: use 'kinds' instead
   */
  kind?: string;
  /** Filter by blockchain address */
  address?: string;
  /** Filter by internal address ID */
  addressId?: string;
  /** Filter by multiple kinds of reservation (takes precedence over 'kind') */
  kinds?: string[];
  /** Base64-encoded cursor for pagination */
  cursorCurrentPage?: string;
  /** Page request direction: FIRST, PREVIOUS, NEXT, or LAST */
  cursorPageRequest?: string;
  /** Page size */
  cursorPageSize?: string;
}

File: taurus-protect-sdk-go/pkg/protect/model/reservation.go
package model

import "time"

// Reservation represents a fund reservation in the system.
type Reservation struct {
	// ID is the unique identifier for the reservation.
	ID string `json:"id"`
	// Kind is the type of reservation.
	Kind string `json:"kind,omitempty"`
	// Comment is an optional comment about the reservation.
	Comment string `json:"comment,omitempty"`
	// AddressID references the internal address stored in protect.
	AddressID string `json:"address_id,omitempty"`
	// Address is the blockchain address string/hash.
	Address string `json:"address,omitempty"`
	// Amount is the reserved amount.
	Amount string `json:"amount,omitempty"`
	// CreationDate is when the reservation was created.
	CreationDate time.Time `json:"creation_date,omitempty"`
	// CurrencyInfo contains information about the currency.
	CurrencyInfo *CurrencyInfo `json:"currency_info,omitempty"`
	// Asset contains information about the asset.
	Asset *Asset `json:"asset,omitempty"`
	// ResourceID is the resource ID associated with the reservation.
	ResourceID string `json:"resource_id,omitempty"`
	// ResourceType is the type of resource associated with the reservation.
	ResourceType string `json:"resource_type,omitempty"`
}

// ReservationUTXO represents a UTXO associated with a reservation.
type ReservationUTXO struct {
	// ID is the unique identifier for the UTXO.
	ID string `json:"id"`
	// Hash is the transaction hash.
	Hash string `json:"hash,omitempty"`
	// OutputIndex is the index of the output in the transaction.
	OutputIndex int64 `json:"output_index,omitempty"`
	// Script is the locking script.
	Script string `json:"script,omitempty"`
	// Value is the UTXO value in satoshis or smallest unit.
	Value string `json:"value,omitempty"`
	// ValueString is the human-readable value.
	ValueString string `json:"value_string,omitempty"`
	// BlockHeight is the block height where the UTXO was created.
	BlockHeight string `json:"block_height,omitempty"`
	// ReservationID is the ID of the reservation associated with this UTXO.
	ReservationID string `json:"reservation_id,omitempty"`
}

// ListReservationsOptions contains options for listing reservations.
type ListReservationsOptions struct {
	// Kinds filters by reservation types (e.g., ["PENDING_REQUEST", "MINIMUM_BALANCE"]).
	// Takes precedence over Kind (deprecated).
	Kinds []string
	// Address filters by blockchain address.
	Address string
	// AddressID filters by internal address ID.
	AddressID string
	// PageSize is the page size: 0 selects DefaultPageSize, above MaxPageSize is an error.
	PageSize int64
	// Cursor is a previous page's Page.NextCursor; the SDK then requests the NEXT page.
	Cursor string
	// CurrentPage and PageRequest (FIRST, PREVIOUS, NEXT, LAST) page by hand; neither can be
	// combined with Cursor.
	CurrentPage string
	PageRequest string
}

// ListReservationsResult contains the result of listing reservations.
type ListReservationsResult struct {
	// Reservations is the list of reservations.
	Reservations []*Reservation `json:"reservations"`
	// Page continues the list: pass Page.NextCursor as the next Cursor until HasMore is false.
	Page CursorPage `json:"page"`
}
